sympy/matrices/random.py

```diff
@@ -0,0 +1,2343 @@
+
+    Alternatively (sometimes more handy),
+    a simple list of eigenvalues would work, too.
+
+    As long as an eigenvalue is repeated in **spec**
+    it mandates the same *Jordan* block until
+    the values changes or the sequence of eigenvalues
+    is interrupted by a **None** entry.
+
+    So the above example is equivalent to **spec** = (1,1,2, None, 2)
+
+    .. math::
+
+    Note, if the length of **spec** (**None** entries excluded)
+    does not meet **dim**, blocks will be choosen randomly.
+
+    To meet **dim** the final block might be truncated.
+
+    Examples
+    ========
+
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import jordan_normal
+    >>> seed(1)
+
+    >>> jordan_normal(3, spec=(1,3))
+    Matrix([
+    [1, 0, 0],
+    [0, 1, 0],
+    [0, 0, 3]])
+
+    >>> jordan_normal(3, spec=(2,2,2))  # no more random
+    Matrix([
+    [2, 1, 0],
+    [0, 2, 1],
+    [0, 0, 2]])
+
+    >>> jordan_normal(6, spec=(2,None,2,2,2,2,0))  # no more random
+    Matrix([
+    [2, 0, 0, 0, 0, 0],
+    [0, 2, 1, 0, 0, 0],
+    [0, 0, 2, 1, 0, 0],
+    [0, 0, 0, 2, 1, 0],
+    [0, 0, 0, 0, 2, 0],
+    [0, 0, 0, 0, 0, 0]])
+
+    >>> # equivalent to
+    >>> jordan_normal(6, spec=((2,1),(2,4),(0,1)))  # no more random
+    Matrix([
+    [2, 0, 0, 0, 0, 0],
+    [0, 2, 1, 0, 0, 0],
+    [0, 0, 2, 1, 0, 0],
+    [0, 0, 0, 2, 1, 0],
+    [0, 0, 0, 0, 2, 0],
+    [0, 0, 0, 0, 0, 0]])
+
+    Parameters
+    ==========
+    dim : integer
+        dimension n of matrix
+    spec : tuple or list of symbols (optional)
+        set of values of which scalars (diagonal entries) are choosen.
+
+        * If **dim** meets the sum of dimensions of **spec**,
+          spec will be the eigenvalues (diagonal entries) as it is.
+        * If **dim** and the sum of dimensions of **spec** differs,
+          the diagonal entries will be choosen randomly from **spec**.
+        * If not given **spec** defaults to $\{ -1, 1 \}$.
+
+    See Also
+    ========
+    diagonal_normal
+
+    """
+
+    spec = spec or _elementary_scalars
+
+    # make spec list of (scalar, size) tuples
+    if spec and spec[0] is None:
+        raise ValueError("spec argument must not start with None")
+
+    block_list = list()
+    val, cnt = None, 0
+    for s in spec:
+        if s == val:
+            cnt += 1
+            continue
+        elif s is None:
+            block_list.append((val, cnt))
+            val, cnt = None, 0
+            continue
+        elif cnt:
+            block_list.append((val, cnt))
+        if isinstance(s, (tuple, list)) and len(s) == 2:
+            block_list.append(s)
+            val, cnt = None, 0
+            continue
+        val, cnt = s, 1
+    if cnt:
+        block_list.append((val, cnt))
+    spec = block_list
+    del block_list
+
+    # choose block randomly if dim and len(spec) does not meet
+    if not dim == sum(cnt for val, cnt in spec):
+        spec = tuple(_sample(spec) for _ in range(dim))
```

Comment:
This signature is a bit strange e.g.:
```python
In [7]: jordan_normal(2, (1,2,3)) # random
Out[7]: 
⎡3  0⎤
⎢    ⎥
⎣0  3⎦

In [8]: jordan_normal(3, (1,2,3)) # not random
Out[8]: 
⎡1  0  0⎤
⎢       ⎥
⎢0  2  0⎥
⎢       ⎥
⎣0  0  3⎦

In [9]: jordan_normal(4, (1,2,3)) # random
Out[9]: 
⎡1  0  0  0⎤
⎢          ⎥
⎢0  1  0  0⎥
⎢          ⎥
⎢0  0  3  0⎥
⎢          ⎥
⎣0  0  0  1⎦
```
Also this seems odd:
```python
In [24]: jordan_normal(4, ((7,2),(9,3)))
Out[24]: 
⎡7  1  0  0⎤
⎢          ⎥
⎢0  7  0  0⎥
⎢          ⎥
⎢0  0  1  0⎥
⎢          ⎥
⎣0  0  0  1⎦
```

Reply:
It feels like maybe there's too much going on in one function here and with a single very overloaded argument.

Reply:
Well, but very useful im practice. Moreover, the `spec` argument is used by `trigonalizable`. More precise it is directly forwarded. Here, I think,
```
trigonalizable(5, spec=[-1, 1, 2])
```
is by far more convenient (and natural) than 
```
trigonalizable(5, spec=[(-1,1), (1,1), (2,1)])
```
which is equivalent.

Nevertheless, at first glance it might be confusing as
```
>>> jordan_normal(3, (1, 1, 1))  # not random
Matrix([
[1, 1, 0],
[0, 1, 1],
[0, 0, 1]])
>>> jordan_normal(3, (1, 1, 1, 1))  # random
Out[5]: 
Matrix([
[1, 0, 0],
[0, 1, 0],
[0, 0, 1]])
```
nevertheless I would recommend to keep it.